File: GOF/Prototype/cpp/maze/enchantedroom.cpp
//
// Created by gaojian on 19-2-27.
//

#include "enchantedroom.h"

using std::make_shared;

EnchantedRoom::EnchantedRoom(int n, std::shared_ptr<Spell> spell)
	: Room{n}
{}

EnchantedRoom::EnchantedRoom(std::shared_ptr<EnchantedRoom> enchantedRoom)
	: Room{enchantedRoom}
{}

std::shared_ptr<MapSite> EnchantedRoom::clone()
{
	return make_shared<EnchantedRoom>(std::dynamic_pointer_cast<EnchantedRoom>(shared_from_this()));
}

File: GOF/Prototype/java/src/maze/EnchantedRoom.java
package maze;

public class EnchantedRoom extends Room {
    Spell m_spell;

    public EnchantedRoom(int n, Spell spell)
    {
        super(n);
        m_spell = spell;
    }

    @Override
    public Object clone() throws CloneNotSupportedException {
        return super.clone();
    }

    void initialize(int n, Spell spell) {
        super.initialize(n);
        m_spell = spell;
    }
}
